### PR TITLE
SCB-138 use grpc bidirectional streaming to connect alpha and omega

### PR DESCRIPTION
**Pull request by eric-lee-ltk**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

File: alpha/alpha-server/src/main/java/io/servicecomb/saga/alpha/server/GrpcOmegaCallback.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  public GrpcOmegaCallback(StreamObserver<GrpcCompensateCommand> observer) {
+    this.observer = observer;
+  }
+
+  @Override
+  public void compensate(TxEvent event) {
+    GrpcCompensateCommand command = GrpcCompensateCommand.newBuilder()
+        .setGlobalTxId(event.globalTxId())
+        .setLocalTxId(event.localTxId())
+        .setParentTxId(event.parentTxId().isEmpty() ? "" : event.parentTxId())
```
Comment (seanyinx) [inline alpha/alpha-server/src/main/java/io/servicecomb/saga/alpha/server/GrpcOmegaCallback.java]: parentTxId will be null, if this sub tx has no parent.

File: alpha/alpha-server/src/main/java/io/servicecomb/saga/alpha/server/GrpcTxEventStreamObserver.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import io.grpc.stub.StreamObserver;
+import io.servicecomb.saga.alpha.core.OmegaCallback;
+import io.servicecomb.saga.alpha.core.TxConsistentService;
+import io.servicecomb.saga.alpha.core.TxEvent;
+import io.servicecomb.saga.pack.contract.grpc.GrpcCompensateCommand;
+import io.servicecomb.saga.pack.contract.grpc.GrpcTxEvent;
+
+class GrpcTxEventStreamObserver implements StreamObserver<GrpcTxEvent> {
+
+  private static Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
```
Comment (seanyinx) [inline alpha/alpha-server/src/main/java/io/servicecomb/saga/alpha/server/GrpcTxEventStreamObserver.java]: final please

File: alpha/alpha-server/src/main/java/io/servicecomb/saga/alpha/server/GrpcTxEventStreamObserver.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    this.omegaCallbacks = omegaCallbacks;
+    this.txConsistentService = txConsistentService;
+    this.responseObserver = responseObserver;
+  }
+
+  @Override
+  public void onNext(GrpcTxEvent message) {
+    // register a callback on started event
+    if (message.getType().equals(TxStartedEvent.name())) {
+      Map<String, OmegaCallback> callbacks = new ConcurrentHashMap<>();
```
Comment (seanyinx) [inline alpha/alpha-server/src/main/java/io/servicecomb/saga/alpha/server/GrpcTxEventStreamObserver.java]: will subsequent TxStartedEvent from the same service but different instance overwrite the map, causing the map having only one instance at most?

File: alpha/alpha-server/src/main/java/io/servicecomb/saga/alpha/server/GrpcTxEventStreamObserver.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        message.getParentTxId().isEmpty() ? null : message.getParentTxId(),
+        message.getType(),
+        message.getCompensationMethod(),
+        message.getPayloads().toByteArray()
+    ));
+  }
+
+  @Override
+  public void onError(Throwable t) {
+    LOG.error("failed to process grpc message.", t);
```
Comment (seanyinx) [inline alpha/alpha-server/src/main/java/io/servicecomb/saga/alpha/server/GrpcTxEventStreamObserver.java]: does onError & onCompleted mean the client is disconnected? if so, we need to find a way to remove the instance from the callback map

File: omega/omega-connector/omega-connector-grpc/src/main/java/io/servicecomb/saga/omega/connector/grpc/GrpcCompensateStreamObserver.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.grpc.stub.StreamObserver;
+import io.servicecomb.saga.omega.transaction.MessageHandler;
+import io.servicecomb.saga.pack.contract.grpc.GrpcCompensateCommand;
+
+public class GrpcCompensateStreamObserver implements StreamObserver<GrpcCompensateCommand> {
+
+  private static Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
```
Comment (seanyinx) [inline omega/omega-connector/omega-connector-grpc/src/main/java/io/servicecomb/saga/omega/connector/grpc/GrpcCompensateStreamObserver.java]: final please

File: alpha/alpha-server/src/test/java/io/servicecomb/saga/alpha/server/AlphaIntegrationTest.java [hunk 6]
```diff
@@ -80,14 +82,32 @@
   @Autowired
   private List<CompensationContext> compensationContexts;
 
+  // use an empty response observer as we don't need the response in client side
+  private final StreamObserver<GrpcCompensateCommand> emptyResponseObserver = new StreamObserver<GrpcCompensateCommand>() {
```
Comment (seanyinx) [inline alpha/alpha-server/src/test/java/io/servicecomb/saga/alpha/server/AlphaIntegrationTest.java]: do we need an integration test in PackIT to verify the compensation callback between alpha and omega?

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/GrpcTxEventStreamObserver.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+  @Override
+  public void onError(Throwable t) {
+    LOG.error("failed to process grpc message.", t);
+    onCompleted();
+  }
+
+  @Override
+  public void onCompleted() {
+    responseObserver.onCompleted();
```
Comment (WillemJiang) [inline alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/GrpcTxEventStreamObserver.java]: It's better to write some log for it.

File: alpha/alpha-server/src/test/java/org/apache/servicecomb/saga/alpha/server/AlphaIntegrationTest.java [hunk 8]
```diff
@@ -95,119 +103,86 @@ public void persistsEvent() throws Exception {
...
-        new CompensationContext(globalTxId, this.localTxId, "method a", "service a".getBytes()),
-        new CompensationContext(globalTxId, localTxId1, "method b", "service b".getBytes())
+    String parentTxId1 = UUID.randomUUID().toString();
+    requestObserver.onNext(eventOf(TxStartedEvent, localTxId1, parentTxId1, "service b".getBytes(), "method b"));
+    requestObserver.onNext(eventOf(TxEndedEvent, new byte[0], "method b"));
+
+    requestObserver.onNext(someGrpcEvent(TxAbortedEvent));
+
+    await().atMost(1, SECONDS).until(() -> receivedCommands.size() > 1);
+    System.out.println(receivedCommands.size());
```
Comment (WillemJiang) [inline alpha/alpha-server/src/test/java/org/apache/servicecomb/saga/alpha/server/AlphaIntegrationTest.java]: Please don't use the System.out here to verify the result.

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/GrpcTxEventStreamObserver.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+  public void onNext(GrpcTxEvent message) {
+    // register a callback on started event
+    String serviceName = message.getServiceName();
+    String instanceId = message.getInstanceId();
+    if (message.getType().equals(TxStartedEvent.name())) {
+      Map<String, OmegaCallback> instanceCallback = omegaCallbacks
+          .computeIfAbsent(serviceName, v -> new ConcurrentHashMap<>());
+      instanceCallback.putIfAbsent(instanceId, new GrpcOmegaCallback(responseObserver));
+      Map<String, String> serviceInstanceId = omegaCallbacksReverse
+          .computeIfAbsent(responseObserver, v -> new ConcurrentHashMap<>());
```
Comment (seanyinx) [inline alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/GrpcTxEventStreamObserver.java]: why using map, if each observer corresponds to only one instance?

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/GrpcOmegaCallback.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        .setLocalTxId(event.localTxId())
+        .setParentTxId(event.parentTxId() == null ? "" : event.parentTxId())
+        .setCompensateMethod(event.compensationMethod())
+        .setPayloads(ByteString.copyFrom(event.payloads()))
+        .build();
+    observer.onNext(command);
+  }
+
+  @Override
+  public boolean equals(Object o) {
```
Comment (seanyinx) [inline alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/GrpcOmegaCallback.java]: why equals & hashcode?

File: alpha/alpha-server/src/test/java/org/apache/servicecomb/saga/alpha/server/AlphaIntegrationTest.java [hunk 11]
```diff
@@ -95,119 +103,84 @@ public void persistsEvent() throws Exception {
...
-        .setCompensationMethod(getClass().getCanonicalName())
-        .setPayloads(ByteString.copyFrom(payload.getBytes()))
-        .build();
+  @Test
+  public void getCompensateCommandOnFailure() {
+    StreamObserver<GrpcTxEvent> requestObserver = stub.callbackCommand(compensateResponseObserver);
+    requestObserver.onNext(someGrpcEvent(TxStartedEvent));
+    await().atMost(1, SECONDS).until(() -> eventRepo.findByEventGlobalTxId(globalTxId) != null);
+    requestObserver.onNext(someGrpcEvent(TxAbortedEvent));
+    await().atMost(1, SECONDS).until(() -> !receivedCommands.isEmpty());
```
Comment (seanyinx) [inline alpha/alpha-server/src/test/java/org/apache/servicecomb/saga/alpha/server/AlphaIntegrationTest.java]: please add blank line between given/when/then to improve readability

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/GrpcTxEventEndpointImpl.java [hunk 12]
```diff
@@ -20,39 +20,35 @@
...
+import org.apache.servicecomb.saga.alpha.core.OmegaCallback;
 import org.apache.servicecomb.saga.alpha.core.TxConsistentService;
-import org.apache.servicecomb.saga.alpha.core.TxEvent;
-import org.apache.servicecomb.saga.pack.contract.grpc.GrpcEmpty;
+import org.apache.servicecomb.saga.pack.contract.grpc.GrpcCompensateCommand;
 import org.apache.servicecomb.saga.pack.contract.grpc.GrpcTxEvent;
 import org.apache.servicecomb.saga.pack.contract.grpc.TxEventServiceGrpc.TxEventServiceImplBase;
 
+import io.grpc.stub.StreamObserver;
+import javafx.util.Pair;
```
Comment (seanyinx) [inline alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/GrpcTxEventEndpointImpl.java]: javafx is for GUI....... we don't want to involve the lib here

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/GrpcTxEventStreamObserver.java [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+import io.grpc.stub.StreamObserver;
+import javafx.util.Pair;
+
+class GrpcTxEventStreamObserver implements StreamObserver<GrpcTxEvent> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  private final Map<String, Map<String, OmegaCallback>> omegaCallbacks;
+
+  private final Map<StreamObserver<GrpcCompensateCommand>, Pair<String, String>> omegaCallbacksReverse;
```
Comment (seanyinx) [inline alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/GrpcTxEventStreamObserver.java]: why using map, if the key is not needed?

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/AlphaConfig.java [hunk 14]
```diff
@@ -42,37 +45,47 @@
...
 
-  // TODO: 2017/12/27 to be filled with actual callbacks on completion of SCB-138
+  // TODO: 2018/01/03 optimize reverse visit of the map instead of using another map, namely omegaCallbacksReverse
   @Bean
   Map<String, Map<String, OmegaCallback>> omegaCallbacks() {
     return new ConcurrentHashMap<>();
   }
 
+  @Bean
+  Map<StreamObserver<GrpcCompensateCommand>, Pair<String, String>> omegaCallbacksReverse() {
```
Comment (seanyinx) [inline alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/AlphaConfig.java]: we use a shared omega callbacks map, in order to use it in service/instance aware callback. why do we need a shared map here?

**Pull request merged**